docs/rfcs/0001-integration-with-dune.md

```diff
@@ -0,0 +1,169 @@
+- Start Date: 2021-05-31
+- RFC PR: https://github.com/melange-re/melange/pull/200
+
+# Summary
+
+To allow Melange users to leverage Dune `library` and `executable` stanzas as
+it is currently possible with any other OCaml project.
+
+# Basic example
+
+Dune would allow to use Melange as another cross-compiler, e.g.
+
+```dune
+(context (default (targets (melange))))
+```
+
+Then, users would be able to define `library` and `executable` stanzas that
+would be built by calling Melange binary `ocamlc`. To define the output format
+and extension, a new stanza `melange` would be supported in `dune` files:
+```dune
+(executable
+    (name foo)
+    (modes melange)
+    (melange (module es6) (suffix bs.js))
+)
+```
+
+# Motivation
+
+The main motivations are:
+
+- Improve integration with existing projects with OCaml and Melange sources: by
+adding first-class support for Melange in Dune, it becomes easier to configure
+projects that share libraries between OCaml native and Melange targets.
+- Improve build performance of Melange projects: the current way Melange defines
+all rules explicitly adds some overhead to build process, which could be removed
+if Melange was exposed as any other OCaml compiler. 
+
+# Detailed design
+
+Unlike Js_of_ocaml, `.cmo` artifacts generated by Melange can't be shared with
+the default `native` environment, as Melange does not transform bytecode, but
+starts to translate to JavaScript from [the lambda representation](https://dev.realworldocaml.org/compiler-backend.html).
+So the least intrusive way to model Melange and Dune integration is to leverage
+Dune's existing support for cross-compilation.
+
+Changes required in Melange:
+
+- expose `bsc` as `ocamlc`
+- generate `cmo` files instead of `cmj` files
```

Comment:
As we're introducing this new mode, such thing is not needed and it's better to have our own file names

Reply:
I see. I thought reusing `cmo` would require less changes in Dune, if we are going to leverage cross compilation in any case. But if this will make things more confusing, I guess we can generate both (although it'd be a bit wasteful from perf perspective?)

Reply:
They contain different things so there's not much that can be reused anyway. Adding `cmj` support shouldn't be too hard. We keep extensions fairly centralized in dune.